### ISSUE TITLE
How to know about the length of the data, TF2.0?

### ISSUE THREAD

**Issue by rishabhsahrawat** [tensorflow/tensorflow#30529]
I have been following these two tutorials, [1](https://www.tensorflow.org/beta/tutorials/load_data/text#split_the_dataset_into_text_and_train_batches) & [2](https://www.tensorflow.org/beta/tutorials/load_data/csv) in order to read a csv file to train a NN model. However, In [tutorial 1](https://www.tensorflow.org/beta/tutorials/load_data/text#split_the_dataset_into_text_and_train_batches), [here exactly](https://www.tensorflow.org/beta/tutorials/load_data/text#load_text_into_datasets), I did not understand how the author chose the values for **BUFFFER** and **TAKE**. It is not documented how to find the total number of examples/elements in the txt files because some file might consisting less or more examples/elements. 
Thank you!

**Comment by pandrey-fr** [comment]
This would indeed be a useful feature (I do not believe that it is implemented as of now).

That being said, in most cases you have some knowledge about your data and set up those constants accordingly... You may however run independent python operations to gather files' lengths (in this tutorial, you simply need to sum the number of rows in each file, which is easily obtained - see below) or, provided you are using Eager execution, do `n_samples = sum(1 for _ in iter(all_labeled_data))`.
To count rows (which works when you do not have Eager execution on):
```
n_samples = 0
for file_name in FILE_NAME:
    with open(os.path.join(parent_dir, file_name)) as file:
        n_samples += sum(1 for _ in file)
```
I hope this helps :)

**Comment by rishabhsahrawat** [comment]
Hi Paul, 
thank you for this suggestion. This is the problem that one has to do it separately, to look into my file (in my case, a csv file) I had to execute a separate script consisting Pandas DF. Neat and clean but there is nothing in TF(2.0). I request Tensorflow's team to provide such feature, if possible.

**Comment by pandrey-fr** [comment]
Hi,
I get that, and personally agree that it would be a useful feature. Let's hope @jvishnuvardhan (and/or other TF people) can make it happen!
FYI, you can count rows in a csv files using the same trick as I posted yesterday (merely opening files with `open` and counting how much rows there are, possibly substracting 1 if you have a header), so as not to depend on pandas (and also not to load the data in memory with it, which is RAM-costly) :)

**Comment by rishabhsahrawat** [comment]
I did not know about the not loading data into memory. Thank you, I will try this inexpensive way. :)

**Comment by pandrey-fr** [comment]
To be more precise, `open` creates a generator that yields the data as it is read. By iterating over the resulting object, you load the data in memory, but only one row at a time, which as you do not attempt to keep it further is also discarded between iterations (well, when the garbage collector passes, which may not be at _every_ iteration, but you get idea). Thus, instead of loading all of the file's content (plus parsing it) with pandas, you iteratively load and discard small chunks (the rows) and only keep track of how many times you were able to do that (thus how many rows there are in the csv file, including the header if any, hence my mentioning substracting one).

**Comment by rishabhsahrawat** [comment]
thank you for the nice explanation. :)👍

**Comment by jsimsa** [comment]
Hi, tf.data API provides the [cardinality](https://www.tensorflow.org/api_docs/python/tf/data/experimental/cardinality) operation which returns the size of a tf.data dataset. The caveat is that the operation can return "unknown" if it cannot be done in a constant time with respect to the number of elements of the dataset. In particular, the cardinality operation will not attempt to enumerate all elements stored in a file.

For an arbitrary (finite) `dataset`, you can always count the number of its elements as follows:

``` 
num_elements = 0
for element in dataset:
  num_elements += 1
```

**Comment by pandrey-fr** [comment]
@jsimsa Thank you for pointing `tf.data.experimental.cardinality` out!

(note: the second solution is similar to the `n_samples = sum(1 for _ in iter(all_labeled_data))` line I pointed out in my initial answer - and it requires Eager execution, which is becoming the norm in tf 2.0)

**Comment by rishabhsahrawat** [comment]
Hi @jsimsa , thank you for this information. I tried [cardinality](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/data/experimental/cardinality) but its output is `tf.Tensor(-2, shape=(), dtype=int64)` on my dataset consisting two columns, **Descritption** and **Label**. I am not sure how to get length from this output. However, 

> num_elements = 0
for element in dataset:
      num_elements += 1

behaves as expected. 
Also, Since my dataset has many classes (approx. 100000), I sometimes want to test it on lets' say first 50000 classes. This dataset is sorted on the basis of classes like from class 0,1, 2...100000. Now I want to know the number of elements until class number 50000. Is there a way in TF 2.0, to count elements only until class 50000?. 
I can do it using Pandas DF, where I just select classes until 50000, something like `dataset[dataset['label']<51000]`

**Comment by jsimsa** [comment]
@rishabhsahrawat can you share the code for which you receive `-2` (which by the way is a constant used to identify unknown cardinality)

Unlike Pandas DF, tf.data provide a streaming abstraction and thus supports datasets that do not necessarily fit into memory. If you would like to limit the dataset to its first `X` elements, you can do so through the `take` transformation. If you would like to filter a subset of a dataset based on values of each element, you can do so through the `filter` transformation.

**Comment by rishabhsahrawat** [comment]
Hi, here is the code.
`dataset = tf.data.experimental.make_csv_dataset('name_of_file.csv', batch_size = 1, select_columns = ['description', 'Label'], label_name = 'Label', num_epochs = 1 ,shuffle = True)`
`print 'Cardinality: ',tf.data.experimental.cardinality(dataset)`
As I said, this csv file has two columns, **description** and **Label**.  Total number of elements are **1265814**. 
If you need any other information, please let me know.

thank you for the information about `tf.filter`, I will test it out.

**Comment by jsimsa** [comment]
Your dataset is read from a file. As per my original response, "the `cardinality` operation will not attempt to enumerate all elements stored in a file."

**Comment by rishabhsahrawat** [comment]
okay, I see. So, there is no other way to find the number of elements in the file except the `for` loop you mentioned earlier.?. Also then how and where to use `cardinality`?

**Comment by jsimsa** [comment]
Correct. `cardinality` works with datasets created through `from_tensors` or `from_tensor_slices`.

**Comment by rishabhsahrawat** [comment]
Okay, clear. 
I am playing with `filter`, mentioned earlier by you, does it also work on datasets created through `from_tensors` or `from_tensor_slices`? Because in the example [here](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/data/Dataset#filter), the dataset is created using `from_tensor_slices`.
